ArcGISRuntimeSDKQt_CppSamples/Analysis/ConfigureSubnetworkTrace/ConfigureSubnetworkTrace.cpp

```diff
@@ -0,0 +1,368 @@
+
+    m_traceConfiguration->traversability()->setBarriers(combineExpressions);
+  }
+  else
+  {
+    m_expressionBuilder = expressionToString(expression);
+    emit expressionBuilderChanged();
+
+    m_traceConfiguration->traversability()->setBarriers(expression);
+  }
+}
+
+void ConfigureSubnetworkTrace::changeIncludeBarriersState(bool includeBarriers)
+{
+  m_traceConfiguration->setIncludeBarriers(includeBarriers);
+}
+
+void ConfigureSubnetworkTrace::changeIncludeContainersState(bool includeContainers)
+{
+  m_traceConfiguration->setIncludeContainers(includeContainers);
+}
+
+void ConfigureSubnetworkTrace::reset()
+{
+  // Reset the barrier condition to the initial value.
+  m_traceConfiguration->traversability()->setBarriers(m_initialExpression);
+  m_expressionBuilder.clear();
+  m_expressionBuilder = expressionToString(static_cast<UtilityTraceConditionalExpression*>(m_initialExpression));
+  emit expressionBuilderChanged();
+}
+
+void ConfigureSubnetworkTrace::trace()
+{
+  m_busy = true;
+  emit busyChanged();
+  if (!m_utilityNetwork || !m_utilityElementStartingLocation)
+  {
+    return;
+  }
+  else
+  {
+    const QList<UtilityElement*> startingLocations {m_utilityElementStartingLocation};
+    // Create utility trace parameters for the starting location.
+    m_traceParams = new UtilityTraceParameters(UtilityTraceType::Subnetwork, startingLocations, this);
+    m_traceParams->setTraceConfiguration(m_traceConfiguration);
+
+    // trace the network
+    m_utilityNetwork->trace(m_traceParams);
+  }
+}
+
+void ConfigureSubnetworkTrace::onTraceCompleted()
+{
+  if (m_utilityNetwork->traceResult()->isEmpty())
+  {
+    m_dialogText = QString("No results returned");
+    emit dialogTextChanged();
+    m_dialogVisible = true;
+    emit dialogVisibleChanged();
+  }
+  // Get the first result.
+  UtilityTraceResult* result = m_utilityNetwork->traceResult()->at(0);
+
+  const QList<UtilityElement*> elements = static_cast<UtilityElementTraceResult*>(result)->elements(this);
+
+  // Display the number of elements found by the trace.
+  m_dialogText = QString("%1 elements found.").arg(elements.length());
+  m_dialogVisible = true;
+  m_busy = false;
+  emit dialogTextChanged();
+  emit dialogVisibleChanged();
+  emit busyChanged();
+}
+
+void ConfigureSubnetworkTrace::onUtilityNetworkLoaded(const Error& e)
+{
+  if (!e.isEmpty())
+  {
+    m_dialogText = QString("%1 - %2").arg(e.message(), e.additionalMessage());
+    m_dialogVisible = true;
+    emit dialogTextChanged();
+    emit dialogVisibleChanged();
+    return;
+  }
+
+  m_busy = false;
+  emit busyChanged();
+  m_networkDefinition = m_utilityNetwork->definition();
+
+  // Build the choice lists for network attribute comparison.
+  for (UtilityNetworkAttribute* networkAttribute : m_networkDefinition->networkAttributes())
+  {
+    if (!networkAttribute->isSystemDefined())
+      m_attributeListModel.append(networkAttribute->name());
+  }
+  emit attributeListModelChanged();
```

Comment:
> Shouldn't this be right beneath `m_attributeListModel.append(networkAttribute->name());`?

I only emit the signal once I know the list is done being populated rather than every time one is added.

> What if all of the network attributes are system defined?

In the case of this sample we know that we want to use the attributes that are not system defined. If the user inputs their own data we expect them to make the other necessary changes as well.

Specifically for this sample though, system defined attributes could work but don't make a whole lot of sense in the context of the sample. E.g Tier name or Source ID.